modules/core/src/main/java/org/apache/synapse/message/store/impl/rabbitmq/RabbitMQConsumer.java

```diff
@@ -34,255 +34,220 @@
-				try {
-					storableMessage = (StorableMessage) in.readObject();
-				} catch (ClassNotFoundException e) {
-					logger.error(getId() + "unable to read the stored message" + e);
-					channel.basicAck(delivery.getEnvelope().getDeliveryTag(), false);
-				}
+    private static final Log log = LogFactory.getLog(RabbitMQConsumer.class.getName());
+
+    private Connection connection;
+    private Channel channel;
+    private RabbitMQStore store;
+    private String queueName;
+    private String idString;
+    private CachedMessage cachedMessage; // Holds the last message read from the message store
+
+    public RabbitMQConsumer(RabbitMQStore store) {
+        if (store == null) {
+            log.error("Cannot initialize.");
+            return;
+        }
+        this.store = store;
+        cachedMessage = new CachedMessage();
+    }
 
-				bis.close();
-				in.close();
-				org.apache.axis2.context.MessageContext axis2Mc = store.newAxis2Mc();
-				MessageContext synapseMc = store.newSynapseMc(axis2Mc);
-				synapseMc =
-						MessageConverter.toMessageContext(storableMessage, axis2Mc, synapseMc);
-				updateCache(delivery, synapseMc, null, false);
-				if (logger.isDebugEnabled()) {
-					logger.debug(getId() + " Received MessageId:" +
-					             delivery.getProps().getMessageId());
-				}
-				return synapseMc;
-			}
-		} catch (ShutdownSignalException sse) {
-			logger.error(getId() + " connection error when receiving messages" + sse);
-		} catch (IOException ioe) {
-			logger.error(getId() + " connection error when receiving messages" + ioe);
-		}
-		return null;
-	}
+    /**
+     * Get a single message from the queue and deserialize for set into the message context
+     *
+     * @return the {@link MessageContext} with received message
+     */
+    @Override
+    public MessageContext receive() {
+        if (isAlive()) {
+            try {
+                GetResponse delivery = channel.basicGet(queueName, false);
+                if (delivery != null) {
+                    StorableMessage storableMessage = deserializeMessage(channel, delivery);
+                    org.apache.axis2.context.MessageContext axis2Mc = store.newAxis2Mc();
+                    MessageContext synapseMc = store.newSynapseMc(axis2Mc);
+                    synapseMc = MessageConverter.toMessageContext(storableMessage, axis2Mc, synapseMc);
+                    updateCache(delivery, delivery.getProps().getMessageId());
+                    log.info(getId() + " Received MessageId: " + delivery.getProps().getMessageId());
+                    if (log.isDebugEnabled()) {
+                        log.debug(getId() + " Received MessageId: " + delivery.getProps().getMessageId());
+                    }
+                    return synapseMc;
+                }
+            } catch (ShutdownSignalException | IOException e) {
+                log.error(getId() + " connection error when receiving messages." + e);
```

Comment:
Should not we cleanup the connection and channel upon this? This can be tricky. However, during half-open connection situation or due to any other undesirable network error, we may get an exception while receiving a message. Upon that I think we better reset.

Reply:
Ack.